### PR TITLE
Clean up Prometheus metric collection code

### PR DESCRIPTION
**Pull request by trustin**
- Rename PrometheusMetricCollection(Client|Service) to
  PrometheusMetricCollecting(Client|Service) for consistency
- Remove unnecessary U+2028 unicode character in Javadoc comments
- Fix indentation and markup in MetricLabel example in its Javadoc
- Add `@FunctionalInterface` to MetricLabel
- Use more strict type parameter for MetricLabel
- Move PrometheusRegistry to the internal.metric package
- Revamp PrometheusRegistry so that a user doesn't have to know about it
- Overall clean-up of the integration test
  - Move the integration test from the internal.metrics to internal.metric
    (Note the 's')
  - Simplify the regular expressions
  - Reorder the members for readability
- Do not use Javadoc block (/** .. */) for license headers

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/client/metric/PrometheusMetricCollectingClient.java [hunk 1]
```diff
@@ -1,4 +1,4 @@
-/**
+/*
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/metric/PrometheusMetricCollectingClient.java]: @jmostella Seems like you need to fix your IDE settings.

File: core/src/main/java/com/linecorp/armeria/client/metric/PrometheusMetricCollectingClient.java [hunk 2]
```diff
@@ -37,11 +37,11 @@
...
  * @param <O> {@link Response} type
  */
-public final class PrometheusMetricCollectionClient<T extends MetricLabel<T>,
+public final class PrometheusMetricCollectingClient<T extends MetricLabel<T>,
         I extends Request, O extends Response> extends SimpleDecoratingClient<I, O> {
 
     /**
-     * Returns a new {@link Client} decorator that tracks request stats using the Prometheus metrics
 library.
+     * Returns a new {@link Client} decorator that tracks request stats using the Prometheus metrics library.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/metric/PrometheusMetricCollectingClient.java]: There was a hidden unicode character between 'metrics' and 'library'.

File: core/src/main/java/com/linecorp/armeria/server/http/metric/PrometheusExporterHttpService.java [hunk 3]
```diff
@@ -36,23 +37,23 @@
...
+        this.collectorRegistry = collectorRegistry;
     }
 
     @Override
     protected void doGet(ServiceRequestContext ctx, HttpRequest req,
                          HttpResponseWriter res) throws Exception {
         ByteArrayOutputStream stream = new ByteArrayOutputStream();
         try (OutputStreamWriter writer = new OutputStreamWriter(stream)) {
-            TextFormat.write004(writer, prometheusRegistry.metricFamilySamples());
+            TextFormat.write004(writer, collectorRegistry.metricFamilySamples());
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/http/metric/PrometheusExporterHttpService.java]: @kojilin @jmostella Do I need to sort the samples returned by `metricFamilySamples()` by their names?

File: core/src/main/java/com/linecorp/armeria/client/metric/PrometheusMetricCollectingClient.java (hunk 1 above)
Comment (jmostella) [inline core/src/main/java/com/linecorp/armeria/client/metric/PrometheusMetricCollectingClient.java]: That was probably from the left over from our internal codestyle settings

File: core/src/main/java/com/linecorp/armeria/internal/metric/PrometheusRegistry.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+                    if (actualName == null) {
+                        continue;
+                    }
+
+                    actualNames.add(actualName);
+                    containsMatchingName = actualName.equals(name);
+                }
+
+                // Make sure the collector actually produced the metric with the specified metric family name.
+                if (!containsMatchingName) {
```
Comment (hyangtack) [inline core/src/main/java/com/linecorp/armeria/internal/metric/PrometheusRegistry.java]: I'm not sure, but this value depends on the last call of `containsMatchingName = actualName.equals(name)` of line 107? Is it correct?

Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/internal/metric/PrometheusRegistry.java]: You are right. It has to be set back to false again. Will fix.

Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/internal/metric/PrometheusRegistry.java]: Fixed and added the test case methods.

File: core/src/main/java/com/linecorp/armeria/internal/metric/PrometheusRegistry.java [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    static <T extends Collector> T registerIfAbsent(
+            CollectorRegistry collectorRegistry, String name, Function<String, T> collectorFactory) {
+        requireNonNull(collectorRegistry, "collectorRegistry");
+        requireNonNull(name, "name");
+        requireNonNull(collectorFactory, "collectorFactory");
+
+        checkArgument(METRIC_NAME_PATTERN.matcher(name).matches(),
+                      "name: %s (expected: matches %s)", name, METRIC_NAME_PATTERN);
+
+        synchronized (collectors) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/internal/metric/PrometheusRegistry.java]: Sounds good. Fixed.

File: core/src/main/java/com/linecorp/armeria/server/http/metric/PrometheusExporterHttpService.java (hunk 3 above)
Comment (jmostella) [inline core/src/main/java/com/linecorp/armeria/server/http/metric/PrometheusExporterHttpService.java]: I think this line is ok. Now that I remember I think it was a missing ``HELP`` and ``TYPE`` messages, when I tried writing my own formatter before. apparently they are used as delimiters between metrics, down stream.

```
# HELP executors_repetitively Generated from dropwizard metric import
# TYPE executors_repetitively counter
executors_repetitively{name="redis-cache-check-timeout-thread.scheduled",} 0.0
# HELP executors_overrun Generated from dropwizard metric import
# TYPE executors_overrun counter
executors_overrun{name="redis-cache-check-timeout-thread.scheduled",} 0.0
# HELP armeria_client_circuit_breaker_counter_count_success Generated from dropwizard metric import
# TYPE armeria_client_circuit_breaker_counter_count_success gauge
armeria_client_circuit_breaker_counter_count_success{handler="PersonaApi",} 0.0
armeria_client_circuit_breaker_counter_count_success{handler="ShopLoggingService",} 1.0
# HELP armeria_client_circuit_breaker_counter_count_failure Generated from dropwizard metric import
# TYPE armeria_client_circuit_breaker_counter_count_failure gauge
armeria_client_circuit_breaker_counter_count_failure{handler="PersonaApi",} 0.0
armeria_client_circuit_breaker_counter_count_failure{handler="ShopLoggingService",} 0.0
# HELP cache_evictions Generated from dropwizard metric import
# TYPE cache_evictions counter
cache_evictions{local="true",name="Stickershop.getRecommendationsForPackage",} 79.0
cache_evictions{local="true",name="Stickershop.getRecommendationsForUser",} 125.0
cache_evictions{local="true",name="authorStorage.getAuthor",} 320.0
cache_evictions{local="true",name="editorsPickService.getAvailableEditorsPick",} 1.0
cache_evictions{local="true",name="editorsPickService.getAvailableEditorsPicks",} 75.0
cache_evictions{local="true",name="personaAPIService.getUserProfile",} 612.0
cache_evictions{local="true",name="productStorage.getProduct",} 0.0
cache_evictions{local="true",name="productStorage.getProductByVersion",} 690.0
cache_evictions{local="true",name="recommendationDataStorage.getRecommendationForProductDataList",} 32.0
cache_evictions{local="true",name="recommendationMetaStorage.getRecommendationMeta",} 19.0
```
But you should be ok with just using ``TextFormat.write004``

File: core/src/test/java/com/linecorp/armeria/internal/metric/PrometheusRegistryTest.java [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+import com.google.common.collect.ImmutableList;
+
+import io.prometheus.client.Collector;
+import io.prometheus.client.CollectorRegistry;
+import io.prometheus.client.Counter;
+import io.prometheus.client.SimpleCollector;
+
+public class PrometheusRegistryTest {
+
```
Comment (jmostella) [inline core/src/test/java/com/linecorp/armeria/internal/metric/PrometheusRegistryTest.java]: can remove blank line

File: core/src/test/java/com/linecorp/armeria/internal/metric/PrometheusRegistryTest.java [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+        final Collector c = registerIfAbsent(registry, name,
+                                             n -> new CollectorWithMultipleNames(name2, name, name3));
+
+        assertThat(collectors.get(registry)).containsEntry(name, c)
+                                            .containsEntry(name2, c)
+                                            .containsEntry(name3, c);
+    }
+
+    private static final class CollectorWithMultipleNames extends SimpleCollector<Object> {
+
```
Comment (jmostella) [inline core/src/test/java/com/linecorp/armeria/internal/metric/PrometheusRegistryTest.java]: can remove blank line

File: core/src/test/java/com/linecorp/armeria/internal/metric/PrometheusRegistryTest.java [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+                    name(names[0]);
+                    help(HELP);
+                }
+
+                @Override
+                public CollectorWithMultipleNames create() {
+                    throw new UnsupportedOperationException();
+                }
+            });
+
```
Comment (jmostella) [inline core/src/test/java/com/linecorp/armeria/internal/metric/PrometheusRegistryTest.java]: can remove blank line

File: thrift/src/test/java/com/linecorp/armeria/it/metric/PrometheusMetricsIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+    };
+
+    @Test
+    public void hello_first_second_endpoint() throws Exception {
+        hello_first_endpoint();
+        hello_second_endpoint();
+    }
+
+    private static void hello_first_endpoint() throws Exception {
+
```
Comment (jmostella) [inline thrift/src/test/java/com/linecorp/armeria/it/metric/PrometheusMetricsIntegrationTest.java]: can remove blank line

File: thrift/src/test/java/com/linecorp/armeria/it/metric/PrometheusMetricsIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+import com.linecorp.armeria.server.http.metric.PrometheusExporterHttpService;
+import com.linecorp.armeria.server.metric.PrometheusMetricCollectingService;
+import com.linecorp.armeria.server.thrift.THttpService;
+import com.linecorp.armeria.service.test.thrift.main.HelloService.Iface;
+import com.linecorp.armeria.testing.server.ServerRule;
+
+import io.prometheus.client.CollectorRegistry;
+
+public class PrometheusMetricsIntegrationTest {
+
```
Comment (jmostella) [inline thrift/src/test/java/com/linecorp/armeria/it/metric/PrometheusMetricsIntegrationTest.java]: can remove blank line

File: core/src/test/java/com/linecorp/armeria/internal/metric/PrometheusRegistryTest.java (hunk 6 above)
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/internal/metric/PrometheusRegistryTest.java]: Intentioanlly left blank for an aesthetic reason.

File: core/src/test/java/com/linecorp/armeria/internal/metric/PrometheusRegistryTest.java (hunk 7 above)
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/internal/metric/PrometheusRegistryTest.java]: Intentioanlly left blank for an aesthetic reason.

File: core/src/test/java/com/linecorp/armeria/internal/metric/PrometheusRegistryTest.java (hunk 8 above)
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/internal/metric/PrometheusRegistryTest.java]: Intentioanlly left blank for an aesthetic reason.

File: thrift/src/test/java/com/linecorp/armeria/it/metric/PrometheusMetricsIntegrationTest.java (hunk 9 above)
Comment (trustin) [inline thrift/src/test/java/com/linecorp/armeria/it/metric/PrometheusMetricsIntegrationTest.java]: Fixed

File: thrift/src/test/java/com/linecorp/armeria/it/metric/PrometheusMetricsIntegrationTest.java (hunk 10 above)
Comment (trustin) [inline thrift/src/test/java/com/linecorp/armeria/it/metric/PrometheusMetricsIntegrationTest.java]: Intentioanlly left blank for an aesthetic reason.

File: core/src/main/java/com/linecorp/armeria/internal/metric/PrometheusRegistry.java [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+                    }
+
+                    actualNames.add(actualName);
+                    if (actualName.equals(name)) {
+                        containsMatchingName = true;
+                    }
+                }
+
+                // Make sure the collector actually produced the metric with the specified metric family name.
+                if (!containsMatchingName) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/internal/metric/PrometheusRegistry.java]: Fixed

File: core/src/main/java/com/linecorp/armeria/server/http/metric/PrometheusExporterHttpService.java (hunk 3 above)
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/http/metric/PrometheusExporterHttpService.java]: Thanks for explanation, @jmostella. Please keep in mind that this is a public place and thus we should not expose our API :-)

Comment (kojilin) [inline core/src/main/java/com/linecorp/armeria/server/http/metric/PrometheusExporterHttpService.java]: No u don't need to sort if u don't export multiple `registry` in one response.

**Comment by trustin** [comment]
@kojilin Please feel free to merge once @jmostella approves.

File: core/src/main/java/com/linecorp/armeria/common/metric/MetricLabel.java [hunk 12]
```diff
@@ -16,21 +16,23 @@
 
 package com.linecorp.armeria.common.metric;
 
+import io.prometheus.client.Collector;
```
Comment (jmostella) [inline core/src/main/java/com/linecorp/armeria/common/metric/MetricLabel.java]: Sorry why import if its only used for javadocs?

File: core/src/main/java/com/linecorp/armeria/server/http/metric/PrometheusExporterHttpService.java (hunk 3 above)
Comment (jmostella) [inline core/src/main/java/com/linecorp/armeria/server/http/metric/PrometheusExporterHttpService.java]: oh oops

redacted

**Comment by jmostella** [comment]
mostly looks good! 

I think some confusion was caused originally by the naming of ``CollectorRegistry``. I don't think that object is actually a registry, maybe it was just miss named?

File: core/src/main/java/com/linecorp/armeria/common/metric/MetricLabel.java (hunk 12 above)
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/metric/MetricLabel.java]: For a shorter `{@link ..}` tag.

**Comment by trustin** [comment]
> maybe it was just miss named?

Yeah, let me rename `PrometheusRegistry` to something else and ping you.

**Comment by trustin** [comment]
@jmostella Renamed `PrometheusRegistry` to `PrometheusUtil` and updated the class description.

**Comment by jmostella** [comment]
@trustin Sorry, actually I meant the one class from the CollectorRegistry from Prometheus client lib. But I think this is a good change as well :)

**Comment by trustin** [comment]
Thanks, @jmostella!

**Pull request merged**